### DISCUSSION TITLE
CMake support as an independent static library

### DISCUSSION THREAD

**Discussion by xiongyu0523** [MaJerle/lwesp#99] [General]
Hello,

Have you considered adding CMake support so that CMake-based projects can very easily integrate lwesp as a submodule (no need to change any thing). 

I'm working on a small demo on raspberry-pi-pico using CMake (see https://github.com/xiongyu0523/rp2040-esp32-azure-iot). I simply add two CMakeLists.txt to lwesp and now it works but still needs some deeper considerations, something in mind:

- lwesp/system includes BSP files that will be provided by the application, it is better to move them out of the system folder?
- OS port may contain both OS porting file and memory porting file, group them together?

**Comment by MaJerle** [comment]
CMake is one of the long-time considerations, but I'm just lacking to actually implement it, and will greatly appreciate inputs from community on this one. Btw, I cannot access to you link (is it private repo?)

So considering the file structure:
* Core system folder shall stay there, indeed when user uses lib, it may modify and leave them somewhere else in their project structure.
* OS port and memory port (`lwesp_mem_` module) may be grouped, but only if user doesn't use built-in memory manager (which is pretty lightweight, but it does its job)

If you decide to prepare CMake structure, I will merge it together to the project. Keep in mind that files in `lwesp/system` are only optional and often act as a template/bsp, hence not mandatory to use files from repo as part of build.

**Comment by xiongyu0523** [comment]
Now it is public repo. 

basically, I treat the system folder as a reference only, copy required files out of the lwesp source tree to build 
os port and mem port are directly built into lwesp lib but the platform driver (which depends on specific bsp/hal) is linked in the final executable target.
